src/Board.cpp

```diff
@@ -0,0 +1,185 @@
+
+bool Board::isCurrentTurnPiece(const Position source) const
+{
+    return !isEmpty(source) && !isRival(source);
+}
+
+bool Board::isWayBloacked(const Position& source, const Position& dest) const
+{
+    int dx = dest.x - source.x;
+    int dy = dest.y - source.y;
+
+    if (dx != 0 && dy != 0 && std::abs(dx) != std::abs(dy)) {
+        return false;
+    }
+
+    int x_move = (dx > 0) ? 1 : ((dx < 0) ? -1 : 0);
+    int y_move = (dy > 0) ? 1 : ((dy < 0) ? -1 : 0);
+    Position currentPostion = source;
+    currentPostion.x += x_move;
+    currentPostion.y += y_move;
+    while (currentPostion.x != dest.x || currentPostion.y != dest.y) {
+        if (!isEmpty(currentPostion)) {
+            return true;
+        }
+        currentPostion.x += x_move;
+        currentPostion.y += y_move;
+    }
+    return false;
+
+}
+
+bool Board::isPossibleChess(const Position& source, const Position& dest)
+{
+    auto temp = std::move(m_board[dest.x][dest.y]);
+    m_board[dest.x][dest.y] = std::move(m_board[source.x][source.y]);
+    if (isKing(dest)) 
+        m_king[m_board[dest.x][dest.y]->getPieceColor()] = dest;
+
+
+    bool UnderAttack = isGameOver();
+    if (isKing(dest)) 
+        m_king[m_board[dest.x][dest.y]->getPieceColor()] = source;
+
+
+    m_board[source.x][source.y] = std::move(m_board[dest.x][dest.y]);
+    m_board[dest.x][dest.y] = std::move(temp);
+    return UnderAttack;
+
+}
+
+bool Board::isKing(const Position& location) const
+{
+    return m_board[location.x][location.y]->getPlayer() == Players::KING;
+}
+
+bool Board::isGameOver() const
+{
+    auto opponent_moves = getOpponentMoves();
+
+    for (int i = 0; i < opponent_moves.size(); ++i) {
+        auto opponent_move = opponent_moves[i];
+        if (m_board[opponent_move.x][opponent_move.y]->isAcceptableMove(opponent_move, m_king[m_currentPlayerIsWhite]) &&
+            !isWayBloacked(opponent_move, m_king[m_currentPlayerIsWhite])) {
+            return true;
+        }
+    }
+
+    return false;
+
+
+}
+
+bool Board::isPawn(const Position& location) const
+{
+    return m_board[location.x][location.y]->getPlayer() == Players::PAWN;
+}
+
+void Board::update(const Position soucre, const Position dest)
+{
+    if (isKing(soucre))
+        m_king[m_board[soucre.x][soucre.y]->getPieceColor()] = dest;
+
+
+    if (isPawn(soucre))
+    {
+        Pawn* thePawn = dynamic_cast<Pawn*>(m_board[soucre.x][soucre.y].get());
+        thePawn -> setFirstMove();
+    }
+    m_board[dest.x][dest.y] = std::move(m_board[soucre.x][soucre.y]);
+    m_board[soucre.x][soucre.y] = nullptr;
+    m_currentPlayerIsWhite = !m_currentPlayerIsWhite;
+}
+
+
+
+std::vector<Position> Board::getOpponentMoves() const
+{
+   std::vector<Position> enemyMoves;
+    for (int x = 0; x < ROWS; x++) {
+        for (int y = 0; y < COLS; y++) {
+            Position currentPos = { x, y };
+            if (!isEmpty(currentPos) && !isCurrentTurnPiece(currentPos)) {
+                enemyMoves.push_back(currentPos);

```

Comment:
Since you're constructing the Position objects inside the loop, you can use emplace_back instead of push_back to construct the objects directly in the vector without unnecessary copies.